### PR TITLE
DC/MLX5: fix in flush resource check

### PR DESCRIPTION
**Pull request by hoopoepg**
- there was issue in ep_flush resource check:
  incorrect sequence of evaluation which caused
  extra arbiter processing
- added progress pending into purge_pending routine

### CODE REVIEW

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 1]
```diff
@@ -1249,6 +1249,8 @@ uct_dc_mlx5_iface_dci_do_pending_wait(ucs_arbiter_t *arbiter,
             return UCS_ARBITER_CB_RESULT_STOP;
         }
         uct_dc_mlx5_iface_dci_alloc(iface, ep);
+    } else {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1252]: maybe use assertv? with object pointers

File: src/uct/ib/dc/dc_mlx5.c [hunk 2]
```diff
@@ -1218,6 +1218,24 @@ static uct_rc_iface_ops_t uct_dc_mlx5_iface_ops = {
...
+    uct_dc_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface,
+                                                uct_dc_mlx5_iface_t);
+    uct_dc_mlx5_ep_t *ep       = ucs_derived_of(tl_ep, uct_dc_mlx5_ep_t);
+    uint8_t dci                = ep->dci;
+    uint8_t pool_idx;
+
+    uct_dc_mlx5_ep_pending_purge(tl_ep, cb, arg);
+    if (dci != UCT_DC_MLX5_EP_NO_DCI) {
+        pool_idx = uct_dc_mlx5_iface_dci_pool_index(iface, dci);
+        uct_dc_mlx5_iface_progress_pending(iface, pool_idx);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1235]: it's not ok to progress pendiong from purge
should happened only from progress context
is this fix critical?

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.c:1235]: yes, pending_purge without progress cause arbiter ordering failure
I can add deferred DCI releasing, but we have to make deep refactoring of uct_dc_mlx5_ep_pending_purge routine to not free DCI

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5.c:1235]: i think just remove this line:
```
    if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
        uct_dc_mlx5_iface_dci_free(iface, ep);
    }
```
and schedule progress function which goes over all allocated dcis, and releases it if it belongs to and ep w/o outstanding and w/o pending

File: src/uct/ib/dc/dc_mlx5.c [hunk 3]
```diff
@@ -1405,11 +1391,33 @@ static UCS_CLASS_INIT_FUNC(uct_dc_mlx5_iface_t, uct_md_h tl_md, uct_worker_h wor
...
+    ucs_assert(!ucs_list_is_empty(&iface->tx.dci_release.list));
+
+    ucs_list_for_each_safe(elem, tmp, &iface->tx.dci_release.list, list) {
+        uct_dc_mlx5_iface_dci_release(iface, elem->dci);
+        ucs_mpool_put(elem);
+    }
+
+    uct_worker_progress_unregister_safe(
+            &iface->super.super.super.super.worker->super,
+            &iface->tx.dci_release.prog_id);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1412]: maybe make it oneshot callback?

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 4]
```diff
@@ -482,7 +485,22 @@ static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
         return;
     }
 
-    uct_dc_mlx5_iface_dci_release(iface, dci_index);
+    if (!sched) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:488]: IMO it should be simpler:
- no need to have release list or elements
- progress just goes over all dci in the iface  and releases those without outstanding/pending
- this way no need changes in dci_free (which increase code size of inlined function)

File: src/uct/ib/dc/dc_mlx5.c [hunk 5]
```diff
@@ -1580,3 +1606,30 @@ void uct_dc_mlx5_iface_set_ep_failed(uct_dc_mlx5_iface_t *iface,
     ep->flags |= UCT_DC_MLX5_EP_FLAG_ERR_HANDLER_INVOKED;
 }
 
+unsigned uct_dc_mlx5_ep_dci_release_progress(void *arg)
+{
+    uct_dc_mlx5_iface_t *iface = arg;
+    uct_dc_mlx5_iface_dci_release_t *elem, *tmp;
+    uint8_t pool_index;
+    uint64_t pool_map;
+
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c:1615]: extra space

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 6]
```diff
@@ -492,8 +510,8 @@ static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
 static UCS_F_ALWAYS_INLINE ucs_status_t
 uct_dc_mlx5_iface_dci_get(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
 {
-    uint8_t pool_index = uct_dc_mlx5_ep_pool_index(ep);
-    ucs_arbiter_t *waitq;
+    uint8_t pool_index   = uct_dc_mlx5_ep_pool_index(ep);
+    ucs_arbiter_t *waitq = uct_dc_mlx5_iface_dci_waitq(iface, pool_index);
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.h:514]: no need to init in-place, because it is not used by random scheme

File: src/uct/ib/dc/dc_mlx5.c (hunk 5 above)
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c:1615]: extra line

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.c]: removed

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 1 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: added assert

File: src/uct/ib/dc/dc_mlx5_ep.h (hunk 4 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: ok, done

**Comment by hoopoepg** [comment]
@brminich @yosefe more comments?

File: src/uct/ib/dc/dc_mlx5.c [hunk 7]
```diff
@@ -1377,6 +1377,8 @@ static UCS_CLASS_INIT_FUNC(uct_dc_mlx5_iface_t, uct_md_h tl_md, uct_worker_h wor
 
     uct_rc_mlx5_iface_common_prepost_recvs(&self->super);
 
+    self->tx.dci_release_prog_id = UCS_CALLBACKQ_ID_NULL;
+
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c:1381]: minor: I'd initialize it together with other `self->tx.` fields on line 1327

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 8]
```diff
@@ -1099,6 +1099,10 @@ UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
     uct_dc_mlx5_ep_fc_cleanup(self);
     uct_dc_mlx5_ep_keepalive_cleanup(self);
 
+    if (self->dci != UCT_DC_MLX5_EP_NO_DCI) {
+        uct_dc_mlx5_iface_dci_free(iface, self);
+    }
+
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.c:1105]: need to do it below the check for `uct_dc_mlx5_iface_is_dci_rand` below. With random policy single dci is shared among several eps

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 9]
```diff
@@ -1243,13 +1247,14 @@ uct_dc_mlx5_iface_dci_do_pending_wait(ucs_arbiter_t *arbiter,
     uint8_t pool_index = uct_dc_mlx5_ep_pool_index(ep);
 
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
+    ucs_assertv(ep->dci == UCT_DC_MLX5_EP_NO_DCI,
+                "EP %p (iface=%p) has DCI=%d while is it scheduled in DCI wait "
+                "queue", ep, iface, ep->dci);
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.c:1252]: maybe also print pool index?

File: src/uct/ib/dc/dc_mlx5.c [hunk 10]
```diff
@@ -1580,3 +1585,37 @@ void uct_dc_mlx5_iface_set_ep_failed(uct_dc_mlx5_iface_t *iface,
...
+        }
+
+        uct_dc_mlx5_iface_dci_free(iface, iface->tx.dcis[dci].ep);
+        if (iface->tx.dcis[dci].ep == NULL) {
+            pool_index = uct_dc_mlx5_iface_dci_pool_index(iface, dci);
+            /* suppress coverity error */
+            ucs_assert(pool_index < (sizeof(pool_map) * 8));
+            pool_map |= UCS_BIT(pool_index);
+        }
+    }
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.c:1613]: why are you freeing all allocated dcis?

File: src/uct/ib/dc/dc_mlx5.c [hunk 11]
```diff
@@ -1580,3 +1585,37 @@ void uct_dc_mlx5_iface_set_ep_failed(uct_dc_mlx5_iface_t *iface,
...
+
+    for (dci = 0; dci < ndci; dci++) {
+        ucs_assert(!uct_dc_mlx5_iface_is_dci_keepalive(iface, dci));
+
+        if (iface->tx.dcis[dci].ep == NULL) {
+            continue;
+        }
+
+        uct_dc_mlx5_iface_dci_free(iface, iface->tx.dcis[dci].ep);
+        if (iface->tx.dcis[dci].ep == NULL) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1607]: maybe dci_free should return 0/1 if released?

File: src/uct/ib/dc/dc_mlx5.c [hunk 12]
```diff
@@ -1580,3 +1585,37 @@ void uct_dc_mlx5_iface_set_ep_failed(uct_dc_mlx5_iface_t *iface,
...
+            pool_index = uct_dc_mlx5_iface_dci_pool_index(iface, dci);
+            /* suppress coverity error */
+            ucs_assert(pool_index < (sizeof(pool_map) * 8));
+            pool_map |= UCS_BIT(pool_index);
+        }
+    }
+
+    ucs_for_each_bit(pool_index, pool_map) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1615]: pls add comment

File: src/uct/ib/dc/dc_mlx5.c [hunk 13]
```diff
@@ -1580,3 +1585,37 @@ void uct_dc_mlx5_iface_set_ep_failed(uct_dc_mlx5_iface_t *iface,
...
+    uint8_t pool_index;
+    uint8_t dci;
+
+    ucs_assert(iface->tx.dci_release_prog_id != UCS_CALLBACKQ_ID_NULL);
+    ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
+
+    for (dci = 0; dci < ndci; dci++) {
+        ucs_assert(!uct_dc_mlx5_iface_is_dci_keepalive(iface, dci));
+
+        if (iface->tx.dcis[dci].ep == NULL) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1602]: if pending q is non empty, also no need to release

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 8 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c:1105]: also, seems we may not progress the arbiter of the pool of this released dci during the dci_release progress

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 14]
```diff
@@ -1243,13 +1247,14 @@ uct_dc_mlx5_iface_dci_do_pending_wait(ucs_arbiter_t *arbiter,
...
 
-    if (ep->dci == UCT_DC_MLX5_EP_NO_DCI) {
-        if (!uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
-            return UCS_ARBITER_CB_RESULT_STOP;
-        }
-        uct_dc_mlx5_iface_dci_alloc(iface, ep);
+    if (!uct_dc_mlx5_iface_dci_can_alloc(iface, pool_index)) {
+        return UCS_ARBITER_CB_RESULT_STOP;
     }
+    uct_dc_mlx5_iface_dci_alloc(iface, ep);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1257]: space line before

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 8 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: but we need to progress pending after releasing this dci

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: ok, added

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 15]
```diff
@@ -1104,8 +1105,17 @@ UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
...
+        (self->dci == UCT_DC_MLX5_EP_NO_DCI)) {
+        return;
+    }
+
+    pool_index = uct_dc_mlx5_iface_dci_pool_index(iface, self->dci);
+    if (uct_dc_mlx5_iface_dci_free(iface, self)) {
+        ucs_assert(self->dci == UCT_DC_MLX5_EP_NO_DCI);
+        if (!(iface->flags & UCT_DC_MLX5_IFACE_FLAG_IN_DCI_PROGRESS)) {
+            uct_dc_mlx5_iface_progress_pending(iface, pool_index);
+        }
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1118]: AS discussed, we should not call progress pending from this context. Pending callbacks should be called only from uct_worker_progress() context.
Can we use the fact that uct_dc_mlx5_ep_pending_purge() is scheduling a progress callback, to eventually progress the arbiter? for example, we could mark dci->ep=(void*)-1 to mean that the dci is still held and should be released by uct_dc_mlx5_ep_dci_release_progress

File: src/uct/ib/dc/dc_mlx5.c [hunk 16]
```diff
@@ -1596,26 +1595,34 @@ unsigned uct_dc_mlx5_ep_dci_release_progress(void *arg)
...
 
-        if (iface->tx.dcis[dci].ep == NULL) {
+        if ((iface->tx.dcis[dci].ep == NULL) ||
+            !ucs_arbiter_group_is_empty(&iface->tx.dcis[dci].ep->arb_group)) {
             continue;
         }
 
-        uct_dc_mlx5_iface_dci_free(iface, iface->tx.dcis[dci].ep);
-        if (iface->tx.dcis[dci].ep == NULL) {
+        if (uct_dc_mlx5_iface_dci_free(iface, iface->tx.dcis[dci].ep)) {
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.c]: fixed

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 17]
```diff
@@ -1104,8 +1105,17 @@ UCS_CLASS_CLEANUP_FUNC(uct_dc_mlx5_ep_t)
         kh_del(uct_dc_mlx5_fc_hash, &iface->tx.fc_hash, it);
     }
 
-    if ((self->dci == UCT_DC_MLX5_EP_NO_DCI) ||
-        uct_dc_mlx5_iface_is_dci_rand(iface)) {
+    if (uct_dc_mlx5_iface_is_dci_rand(iface) ||
+        (self->dci == UCT_DC_MLX5_EP_NO_DCI)) {
+        return;
+    }
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: yes, this evaluation is needed

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 15 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: ok, done: mark dci->ep=(void*)-1 as release candidate

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 18]
```diff
@@ -1248,12 +1254,14 @@ uct_dc_mlx5_iface_dci_do_pending_wait(ucs_arbiter_t *arbiter,
 
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
     ucs_assertv(ep->dci == UCT_DC_MLX5_EP_NO_DCI,
-                "EP %p (iface=%p) has DCI=%d while is it scheduled in DCI wait "
-                "queue", ep, iface, ep->dci);
+                "EP %p (iface=%p) has DCI=%d (pool %d) while it is scheduled "
```
[the remark this answers is not in the document]
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: fixed

File: src/uct/ib/dc/dc_mlx5.c [hunk 19]
```diff
@@ -1595,34 +1595,29 @@ unsigned uct_dc_mlx5_ep_dci_release_progress(void *arg)
...
     ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
 
-    iface->flags |= UCT_DC_MLX5_IFACE_FLAG_IN_DCI_PROGRESS;
-
     for (dci = 0; dci < ndci; dci++) {
         ucs_assert(!uct_dc_mlx5_iface_is_dci_keepalive(iface, dci));
 
-        if ((iface->tx.dcis[dci].ep == NULL) ||
-            !ucs_arbiter_group_is_empty(&iface->tx.dcis[dci].ep->arb_group)) {
+        if (iface->tx.dcis[dci].ep != (void*)-1) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5.c:1601]: pls add some `#define` for this like UCT_DC_IFACE_DCI_EP_DESTROYED

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 20]
```diff
@@ -484,15 +483,25 @@ static inline int uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
...
+static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
+                                              uct_dc_mlx5_ep_t *ep)
+{
+    uint8_t dci_index = ep->dci;
+
+    if (!uct_dc_mlx5_iface_dci_detach(iface, ep)) {
+        return;
+    }
+
+    uct_dc_mlx5_iface_dci_release(iface, dci_index);
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:502]: this adds a small overhead to data path (set ep to -1 then to NULL)

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 21]
```diff
@@ -1398,7 +1410,10 @@ void uct_dc_mlx5_ep_pending_purge(uct_ep_h tl_ep, uct_pending_purge_callback_t c
                             &args);
 
     if (ep->dci != UCT_DC_MLX5_EP_NO_DCI) {
-        uct_dc_mlx5_iface_dci_free(iface, ep);
+        uct_worker_progress_register_safe(
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1413]: need to detach dci?

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 22]
```diff
@@ -479,14 +480,26 @@ static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
...
-
-    iface->tx.dcis[dci_index].ep = NULL;
+    iface->tx.dcis[dci_index].ep = (uct_dc_mlx5_ep_t*)-1;
     ep->dci                      = UCT_DC_MLX5_EP_NO_DCI;
     ep->flags                   &= ~UCT_DC_MLX5_EP_FLAG_TX_WAIT;
+
+    return 1;
+}
+
+static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:493]: always inline

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 23]
```diff
@@ -464,13 +465,13 @@ static inline void uct_dc_mlx5_iface_dci_alloc(uct_dc_mlx5_iface_t *iface, uct_d
     pool->stack_top++;
 }
 
-static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
-                                              uct_dc_mlx5_ep_t *ep)
+static UCS_F_ALWAYS_INLINE int
+uct_dc_mlx5_iface_dci_detach(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:469]: maybe pass "new_ep" param to dci_detach which could be either NULL or UCT_DC_IFACE_DCI_EP_DESTROYED?
this way can make sure that after this function, dci will no longer point to ep

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: done

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 24]
```diff
@@ -385,6 +387,7 @@ uct_dc_mlx5_iface_dci_release(uct_dc_mlx5_iface_t *iface, uint8_t dci_index)
 
     pool->stack_top--;
     pool->stack[pool->stack_top] = dci_index;
+    iface->tx.dcis[dci_index].ep = NULL;
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.h:390]: remove?

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h:390]: not sure it is good idea: uct_dc_mlx5_iface_dci_release is used in several places where dci->ep is not updated

File: src/uct/ib/dc/dc_mlx5.h [hunk 25]
```diff
@@ -73,7 +73,7 @@ typedef enum {
     UCT_DC_MLX5_IFACE_FLAG_KEEPALIVE_FULL_HANDSHAKE = UCS_BIT(1),
 
     /** uidx is set to dci idx */
-    UCT_DC_MLX5_IFACE_FLAG_UIDX                     = UCS_BIT(2)
+    UCT_DC_MLX5_IFACE_FLAG_UIDX                     = UCS_BIT(2),
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5.h:76]: leftover?

**Review by brminich — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5.h (hunk 25 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5.h]: yep, fixed

File: src/uct/ib/dc/dc_mlx5_ep.h (hunk 24 above)
Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.h:390]: which places? it wasn't needed before this PR..

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: dci_put and dci_release_progress

fixed

**Review by yosefe — APPROVED** [review]

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 26]
```diff
@@ -1305,17 +1346,27 @@ uct_dc_mlx5_iface_dci_do_dcs_pending_tx(ucs_arbiter_t *arbiter,
     int is_only                = ucs_arbiter_elem_is_only(elem);
     ucs_arbiter_cb_result_t res;
 
+    ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1349]: can remove it

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 27]
```diff
@@ -1305,17 +1346,27 @@ uct_dc_mlx5_iface_dci_do_dcs_pending_tx(ucs_arbiter_t *arbiter,
...
+    if ((res != UCS_ARBITER_CB_RESULT_REMOVE_ELEM) || !is_only) {
+        return res;
+    }
+
+    /* For dcs* policies release dci if this is the last elem in the group
+     * and the dci has no outstanding operations. For example pending
+     * callback did not send anything. (uct_ep_flush or just return ok)
+     */
+    if (!uct_dc_mlx5_iface_dci_detach(iface, ep,
+                                      UCT_DC_IFACE_DCI_EP_DESTROYED)) {
```
Comment (yosefe) [inline src/uct/ib/dc/dc_mlx5_ep.c:1361]: maybe rename this macro to UCT_DC_IFACE_DCI_EP_DETACHED ?

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 26 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: ok, removed

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 27 above)
Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.c]: renamed to UCT_DC_IFACE_DCI_EP_DETACHED

**Review by yosefe — APPROVED** [review]

**Comment by hoopoepg** [comment]
infra issue

**Comment by hoopoepg** [comment]
bot:pipe:retest

File: src/uct/ib/dc/dc_mlx5_ep.c [hunk 28]
```diff
@@ -1346,8 +1346,6 @@ uct_dc_mlx5_iface_dci_do_dcs_pending_tx(ucs_arbiter_t *arbiter,
     int is_only                = ucs_arbiter_elem_is_only(elem);
     ucs_arbiter_cb_result_t res;
 
-    ucs_assert(!uct_dc_mlx5_iface_is_dci_rand(iface));
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.c]: why removed?

Comment (yosefe) [reply src/uct/ib/dc/dc_mlx5_ep.c]: this function is "uct_dc_mlx5_iface_dci_do_**dcs**_pending_tx" so probably no need to check it again

File: src/uct/ib/dc/dc_mlx5_ep.c (hunk 26 above)
Comment (brminich) [reply src/uct/ib/dc/dc_mlx5_ep.c]: why? imo it is valid assert, which makes sure this handler is not used for random policy

File: src/uct/ib/dc/dc_mlx5_ep.h [hunk 29]
```diff
@@ -479,14 +478,26 @@ static inline void uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface,
...
-    iface->tx.dcis[dci_index].ep = NULL;
     ep->dci                      = UCT_DC_MLX5_EP_NO_DCI;
     ep->flags                   &= ~UCT_DC_MLX5_EP_FLAG_TX_WAIT;
+    iface->tx.dcis[dci_index].ep = new_ep;
+
+    return 1;
+}
+
+static UCS_F_ALWAYS_INLINE void
+uct_dc_mlx5_iface_dci_free(uct_dc_mlx5_iface_t *iface, uct_dc_mlx5_ep_t *ep)
```
Comment (brminich) [inline src/uct/ib/dc/dc_mlx5_ep.h:492]: looks like it is not needed anymore

Comment (hoopoepg) [reply src/uct/ib/dc/dc_mlx5_ep.h]: removed

**Review by yosefe — APPROVED** [review]

**Review by brminich — APPROVED** [review]

**Comment by hoopoepg** [comment]
bot:pipe:retest

**Comment by brminich** [comment]
looks ok to squash

**Comment by hoopoepg** [comment]
@yosefe ok to merge?

**Pull request merged**